jax_am/fem/autodiff_utils.py

```diff
@@ -0,0 +1,207 @@
+        return sol
+
+    @forward_solve.defjvp
+    def forward_solve_jvp(primals, tangents):
+        params, = primals
+        params_dot, = tangents
+        sol = forward_solve(params)
+        sol_dot = implicit_jvp_helper(
+            problem, sol, params, params_dot)
+        return sol.reshape(sol.shape), sol_dot.reshape(sol.shape)
+
+    return forward_solve
+
+
+def implicit_jvp_helper(problem, sol0: Array,
+                        params0: Array, params_dot0: Array) -> Array:
+    """Helper function to compute JVP of FEA forward solve.
+
+    The forward solve is setup such that it can use either
+    PETSc or JAX solvers. However, the backward solve is setup such
+    that only JAX solvers can be used.
+
+    Parameters
+    ----------
+    problem
+        FEA problem object
+    sol0
+        Solution of the forward solve
+    params0
+        Parameters of the forward solve
+    params_dot0
+        Parameters of the backward solve
+
+    Returns
+    -------
+    The output tangents of the forward solve
+    """
+    # create residual function
+
+    def residual(dofs, params):
+        """Function calculates the r(u(p), p) based on weak form.
+        r should be equal to 0 at the solution.
+        """
+        problem.set_params(params)
+        res_fn = problem.compute_residual
+        res_fn = get_flatten_fn(res_fn, problem)
+        res_fn = apply_bc(res_fn, problem)
+        return res_fn(dofs)
+
+    # For time-dependent problems
+    problem.set_params(params0)
+    problem.newton_update(sol0)
+    # Construct terms for JVP calculation
+    partial_fn_of_params = partial(residual, sol0)  # r(u=sol, p)
+
+    def change_arg_order_fn(params, dofs): return residual(dofs, params)
+    partial_fn_of_u = partial(change_arg_order_fn, params0)  # r(u, p=params)
+    # dr/du . x
+    _, jvp_fn = jax.linearize(partial_fn_of_u, sol0)
+
+    def backward_matvec(v): return jvp_fn(v.reshape(-1, problem.vec))
+    # dr/drho . v --> need a negative sign here - will provide later
+    _, backward_rhs = jax.jvp(partial_fn_of_params,
+                              (params0, ), (params_dot0, ))
+    # Call a JAX solver
+    precond_matrix = get_jacobi_precond(jacobi_preconditioner(problem))
+
+    def jax_solver_modified(matvec, v): return jax_solve(problem, matvec,
+                                                         v.reshape(-1), None,
+                                                         False,
+                                                         precond_matrix)
+    chosen_bb_solver = jax_solver_modified
+    # Find adjoint value
+    tangent_out = jax.lax.custom_linear_solve(backward_matvec, -1*backward_rhs,
+                                              chosen_bb_solver,
+                                              transpose_solve=chosen_bb_solver)
+    return tangent_out
+
+
+@jax.custom_jvp
+def jax_array_list_to_numpy_diff(jax_array_list:
+                                 list[Array]) -> onp.ndarray:
```

Comment:
On my machine, it throws an error `TypeError: 'type' object is not subscriptable`

Change `list` to `List` with `from typing import List` resolved the issue.